CreateMessage/handler.ts

```diff
@@ -407,76 +412,76 @@ export function CreateMessageHandler(
       );
 
     // here we create an async Task that processes the request
-    const task =
-      // this is a dummy value, it's just used to set the type of error
-      // see https://github.com/gcanti/fp-ts/issues/528#issuecomment-407749612
-      fromEither<CreateMessageHandlerResponse, boolean>(right(true))
-        .chainSecond(
-          // check whether the client can create a message for the recipient
-          fromEither(
-            canWriteMessage(auth.groups, authorizedRecipients, fiscalCode)
-          )
-        )
-        // Verify if the Service has the required quality to sent message
-        .chain(_ =>
-          disableIncompleteServices &&
-          !incompleteServiceWhitelist.includes(serviceId) &&
-          !authorizedRecipients.has(fiscalCode)
-            ? fromEither(
-                ValidService.decode(userAttributes.service).bimap(
+    return pipe(
+      // check whether the client can create a message for the recipient
+      TE.fromEither(
+        canWriteMessage(auth.groups, authorizedRecipients, fiscalCode)
+      ),
+      // Verify if the Service has the required quality to sent message
+      TE.chain(_ =>
+        disableIncompleteServices &&
+        !incompleteServiceWhitelist.includes(serviceId) &&
+        !authorizedRecipients.has(fiscalCode)
+          ? TE.fromEither(
+              pipe(
+                ValidService.decode(userAttributes.service),
+                E.bimap(
                   _1 => ResponseErrorForbiddenNotAuthorizedForRecipient,
                   _1 => true
                 )
               )
-            : fromEither(right(true))
-        )
-        .chain(_ =>
-          // check whether the client can provide default addresses
-          fromEither(canDefaultAddresses(messagePayload))
-        )
-        .chain(_ =>
-          // check whether the client can ask for payment
-          fromEither(
-            canPaymentAmount(
-              messagePayload.content,
-              service.maxAllowedPaymentAmount
             )
+          : TE.right(true)
+      ),
+      TE.chainW(_ =>
+        // check whether the client can provide default addresses
+        TE.fromEither(canDefaultAddresses(messagePayload))
+      ),
+      TE.chainW(_ =>
+        // check whether the client can ask for payment
+        TE.fromEither(
+          canPaymentAmount(
+            messagePayload.content,
+            service.maxAllowedPaymentAmount
           )
         )
-        .chain(_ =>
-          // create a Message document in the database
-          createMessageDocument(
-            messageId,
-            messageModel,
-            auth.userId,
-            fiscalCode,
-            messagePayload.time_to_live,
-            serviceId
-          )
+      ),
+      TE.chainW(_ =>
```

Comment:
What do you think if we use only one `TE.fromEither`
```typescript
pipe(
      // check whether the client can create a message for the recipient
      canWriteMessage(auth.groups, authorizedRecipients, fiscalCode),
      // Verify if the Service has the required quality to sent message
      E.chain(_ =>
        disableIncompleteServices &&
        !incompleteServiceWhitelist.includes(serviceId) &&
        !authorizedRecipients.has(fiscalCode)
          ? pipe(
              ValidService.decode(userAttributes.service),
              E.bimap(
                _1 => ResponseErrorForbiddenNotAuthorizedForRecipient,
                _1 => true
              )
            )
          : E.right(true)
      ),
      E.chainW(_ =>
        // check whether the client can provide default addresses
        canDefaultAddresses(messagePayload)
      ),
      E.chainW(_ =>
        // check whether the client can ask for payment
        canPaymentAmount(
          messagePayload.content,
          service.maxAllowedPaymentAmount
        )
      ),
      TE.fromEither,
      TE.chainW(_ =>
```

Reply:
This way we are making that step synchronous, right?

Reply:
You are right, all the steps before `TE.taskEither` will be executed on the `pipe` creation and not when the `TaskEither` is executed. IMHO is not a problem if the `pipe` is used inside the execution context. Do you think this pattern could generate unwanted behaviors?

Reply:
The only one I see is not taking full advantages of node event loop